clamp down on warnings in REPL code

SethTisue opened the pull request:
configurable warnings ftw!

attn contributors, more PRs in this vein (and the vein of Som's epic #8890) would be welcome. I just did the projects I happen to be working in at the moment

actually fixing warnings is of course good and I did fix some, but simply silencing existing warnings is okay too, the key thing is to enable `-Werror` by hook or by crook, to set a baseline and discourage backsliding

SethTisue commented on build.sbt:
```diff
@@ -713,36 +713,39 @@ lazy val compiler = configureAsSubproject(project)
+    scalacOptions in Compile ++= Seq("-Xlint", "-Werror", "-unchecked"),
   )
   .settings(
-    connectInput in run := true,
```
no longer necessary here now that we pull in `commonSettings`

SethTisue commented on build.sbt:
```diff
@@ -713,36 +713,39 @@ lazy val compiler = configureAsSubproject(project)
   .settings(disableDocs)
   .settings(skip in publish := true)
-  .settings(scalacOptions in Compile += "-Xlint:-deprecation,-inaccessible,-nonlocal-return,-valpattern,-doc-detached,_")
+  .settings(scalacOptions in Compile ++= Seq("-Xlint", "-Werror", "-unchecked", "-Wconf:cat=deprecation:msg=early.initializers:s"))
```
hard — well, annoying, anyway — to get rid of early initializers, because cake

som-snytt commented on build.sbt:
```diff
@@ -713,36 +713,39 @@ lazy val compiler = configureAsSubproject(project)
   .settings(
     name := "scala-compiler-interactive",
-    description := "Scala Interactive Compiler"
+    description := "Scala Interactive Compiler",
```
`description := "Scala Interactive Compiler [sic]"`

som-snytt commented on build.sbt:
```diff
@@ -713,36 +713,39 @@ lazy val compiler = configureAsSubproject(project)
     name := "scala-compiler-interactive",
-    description := "Scala Interactive Compiler"
+    description := "Scala Interactive Compiler",
+    // TODO remove `-unit-special` once we re-STARR on 2.13.3, because scala/bug#11965
```
This is that thing where they call out joke numbers, right?

I keep thinking "Saturday night special", and realized I don't know how I know that term. Wikipedia offers the synonym, "suicide special".

som-snytt commented on build.sbt:
```diff
@@ -713,36 +713,39 @@ lazy val compiler = configureAsSubproject(project)
     libraryDependencies ++= jlineDeps,
     name := "scala-repl-frontend",
-    scalacOptions in Compile += "-Xlint:-inaccessible,-nonlocal-return,-valpattern,-doc-detached,_",
+    scalacOptions in Compile ++= Seq("-Xlint", "-Werror", "-unchecked"),
```
Noice. On the recent linting PR, I noticed that I had pushed these exclusions down to subprojects and they deserved review.

SethTisue replied:
yeah, I figured. we'll get to it, a project at a time and a flag at a time

som-snytt commented on src/interactive/scala/tools/nsc/interactive/Global.scala:
```diff
@@ -424,7 +424,8 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
         loop = false
-        if (!interruptsEnabled) return
+        // TODO refactor to eliminate breakable/break/return?
+        (if (!interruptsEnabled) return): @nowarn("cat=lint-nonlocal-return")
```
I would not have guessed that one during scaladays open mic session.

som-snytt commented on src/interactive/scala/tools/nsc/interactive/Global.scala:
```diff
@@ -1175,9 +1176,12 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
   }
   object CompletionResult {
+    // outer reference thing (twice) is https://github.com/scala/bug/issues/11534 ?
+    @nowarn("cat=unchecked&msg=The outer reference in this type test cannot be checked at run time")
```
The other lint-related PR applies this in common settings. "Saturday night special."

SethTisue replied:
Yeah and I don't know why it didn't take, here. I'll dig into it, at least a little.

SethTisue replied:
after digging:

the thing you put in common settings was `:ws` (summarize warning). I don't think that's what we want, because we really want to either fix these or silence them, not get "2 unchecked warnings; re-run with -unchecked for details". unchecked warnings can be Very Bad Indeed, so "2 unchecked warnings" is, or should be, an alarming thing to see in the build log

I'm not sure if you really _meant_ `:ws`, or whether you confused **s**ummarize with **s**ilence, but in any case, I'll change it to `:s`. it might be preferable to actually audit the individual sites and `@nowarn` them on a case-by-case basis, but I leave that as “future work”, as the academics say

som-snytt commented on src/repl/scala/tools/nsc/interpreter/Imports.scala:
```diff
@@ -153,14 +153,13 @@ trait Imports {
       accessPath append s".$iw"
     }
-    def addWrapper() {
+    def addWrapper(): Unit =
```
I shall resist checking the sordid history. There is a kind of long tail to these idioms.

som-snytt reviewed (approved):
Thanks! Amazing! Possible duplication but no harm/foul/fowl.

som-snytt commented:
Travis says
```
[error] one error found

[error] (repl / Compile / compileIncremental) Compilation failed
```
but I couldn't read the validate-main output.

som-snytt commented on build.sbt:
```diff
@@ -713,36 +713,39 @@ lazy val compiler = configureAsSubproject(project)
-    description := "Scala Interactive Compiler"
+    description := "Scala Interactive Compiler",
+    // TODO remove `-unit-special` once we re-STARR on 2.13.3, because scala/bug#11965
+    scalacOptions in Compile ++= Seq("-Xlint:-unit-special,_", "-unchecked", "-Werror", "-Wconf:cat=deprecation:msg=early.initializers:s"),
```
`-Weekend-werrior`

Oh, is this the thing where patmat warnings are counted and trigger error result? On another PR, it turned out that `-Werror` means a warning was counted, not emitted as text warning. I wonder if rebase will help. That was for `-nowarn` to mean don't summarize those counts; maybe this is a similar thing.

SethTisue replied:
I've been trying to do something with `-Werror` and "werewolf", but yours is better

Sorry, I don't follow the rest of your comment, I'm confused.

som-snytt replied:
https://github.com/scala/scala/pull/8923 is the PR to make `-nowarn` not say `1 warning` when patmat warning (which is post-typer) escapes `-Wconf` suppression (at typer).

Maybe `-Werror` suffers the same issue. Just guessing.

SethTisue replied:
This PR already does have #8923 in its history.

I still don't follow whether you're giving me review feedback about something I should perhaps change in this PR, or whether your remarks are... just musings about something we should be on the lookout for.

Sorry if I'm being dense.

SethTisue commented:
> [error] (repl / Compile / compileIncremental) Compilation failed

Yeah so what bit me here is that Travis is running with the optimizer enabled, which I wasn't doing in local development, and there are optimizer warnings. after `setupPublishCore` locally, I get the warnings/errors. working on an adjustment

som-snytt commented:
I have a branch from the other day where I turned on optimizer warnings (like the message says) and I realized that what the deer is thinking, as it watches the headlights, is, "Does this matter?"

SethTisue commented:
I made a couple of mistakes that I have now corrected:

* `-unchecked` is the default since 2.13.2, so it doesn't make sense to add it in `scalacOptions`
* I had added `.settings(commonSettings)` in a few places, but that was wrong, those settings were already coming in through `configureAsSubproject`

SethTisue commented:
LOL, `[warn] 482 optimizer warnings; re-run enabling -opt-warnings for details, or try -help`

som-snytt commented:
My previous comment was that I'm looking at optimizer warnings, which include "you wrote inline but not final, so I gave up".

SethTisue commented:
I see in `ScriptCommands.scala` that `setupPublishCore` enables the optimizer buildwide. it probably doesn't make a whole lot of sense to enable it in things like `repl-frontend` (as opposed to library/reflect/compiler), but I don't think I'll tinker with it in this PR, I'll just add something that prevents optimizer warnings from detonating `-Werror`

som-snytt commented:
`-opt-warnings:none`?

SethTisue commented:
I think `-Wconf:cat=optimizer:is` will do the trick. there'll still be a summary line in the CI builds, instead of total silence

SethTisue commented:
@som-snytt if it still LGTY, I'll squash before merge

som-snytt commented:
SLGTM. Travis doesn't lie. "If everyone is unit-special, then no one is!"

som-snytt commented:
I think instead of "merge conflict" we should say "opportunity for collaboration".

Also, FTW is WTF backwards. TFW! FWIW.

SethTisue commented:
> I think instead of "merge conflict" we should say "opportunity for collaboration".

with myself, in this case

SethTisue committed:
clamp down on warnings in REPL code

som-snytt commented:
`[error] (interactive / Compile / compileIncremental) Compilation failed`

SethTisue commented:
bad rebase, now fixed.

SethTisue commented:
ticket for general discussions about doing more PRs in this vein: https://github.com/scala/scala-dev/issues/600

som-snytt commented:
Sorry I missed this one, revealed by `-Wvalue-discard`. I guess I am not Lukas. I see there were some fog-of-war discussions in the review, a peril in larger deltas. Usually, I'm glad to remove extraneous enclosing braces because "we'll notice if it's wrong" and also "someday I'll backport indentation warnings".

![image](https://github.com/scala/scala/assets/369425/31861187-cb70-436c-93f6-f216ba73eb76)
